### PR TITLE
Add experimental trace file field

### PR DESCRIPTION
**Pull request by ijjk**
This adds a new field to our `.nft.json` trace files when enabled behind a feature flag to output the hashes for the traced files. 

Closes NEXT-3232

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7e885b9e7c9db43f4f1ae449b7d78b6f491abfe3

`pnpm test-start test/e2e/app-dir/app-edge-root-layout/index.test.ts`
- app-dir edge runtime root layout > should not emit metadata files into bad paths
- app-dir edge runtime root layout > should mark static contain metadata routes as edge functions

<details>
<summary>Expand output</summary>

  ● app-dir edge runtime root layout › should not emit metadata files into bad paths

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app-dir edge runtime root layout › should mark static contain metadata routes as edge functions

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

`pnpm test-start test/e2e/app-dir/app-basepath/index.test.ts`
- app dir - basepath > should successfully hard navigate from pages -> app
- app dir - basepath > should support `basePath`
- app dir - basepath > should support Link with basePath prefixed
- app dir - basepath > should prefix metadata og image with basePath
- app dir - basepath > should prefix redirect() with basePath
- app dir - basepath > should render usePathname without the basePath
- app dir - basepath > should handle redirect in dynamic in suspense boundary routes with basePath
- app dir - basepath > should only make a single RSC call to the current page (/base/refresh)
- app dir - basepath > should only make a single RSC call to the current page (/base/refresh?foo=bar)

<details>
<summary>Expand output</summary>

  ● app dir - basepath › should successfully hard navigate from pages -> app

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should support `basePath`

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should support Link with basePath prefixed

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should prefix metadata og image with basePath

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should prefix redirect() with basePath

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should render usePathname without the basePath

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should handle redirect in dynamic in suspense boundary routes with basePath

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should only make a single RSC call to the current page (/base/refresh)

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

  ● app dir - basepath › should only make a single RSC call to the current page (/base/refresh?foo=bar)

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: dab85e03eda7077aaba176f6e3c4702b6a1bde20

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/app/flying-shuttle.test.ts` (turbopack)
- should output updated trace files > should have file hashes in trace files

<details>
<summary>Expand output</summary>

  ● should output updated trace files › should have file hashes in trace files

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: a9896a5740d9750e043e10a4ea239def02e46b5f

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/app/flying-shuttle.test.ts` (turbopack)
- should output updated trace files > should have file hashes in trace files

<details>
<summary>Expand output</summary>

  ● should output updated trace files › should have file hashes in trace files

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| buildDuration | 18.8s | 17.4s | N/A |
| buildDurationCached | 9.5s | 8.2s | N/A |
| nodeModulesSize | 360 MB | 360 MB | ⚠️ +61.6 kB |
| nextStartRea..uration (ms) | 457ms | 473ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| 139-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| 2478adb3-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 4967-HASH.js gzip | 5.1 kB | 5.1 kB | N/A |
| 6701.HASH.js gzip | 168 B | 168 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 229 B | 227 B | N/A |
| main-HASH.js gzip | 31.6 kB | 31.6 kB | N/A |
| webpack-HASH.js gzip | 1.65 kB | 1.64 kB | N/A |
| Overall change | 45.3 kB | 45.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 193 B | N/A |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 510 B | 510 B | ✓ |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 2.51 kB | 2.52 kB | N/A |
| edge-ssr-HASH.js gzip | 266 B | 265 B | N/A |
| head-HASH.js gzip | 365 B | 362 B | N/A |
| hooks-HASH.js gzip | 391 B | 390 B | N/A |
| image-HASH.js gzip | 4.32 kB | 4.32 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 393 B | 395 B | N/A |
| withRouter-HASH.js gzip | 323 B | 323 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.73 kB | 1.73 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 484 B | ✓ |
| Overall change | 484 B | 484 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| index.html gzip | 527 B | 528 B | N/A |
| link.html gzip | 540 B | 541 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| edge-ssr.js gzip | 108 kB | 108 kB | N/A |
| page.js gzip | 3.04 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 624 B | N/A |
| middleware-r..fest.js gzip | 155 B | 154 B | N/A |
| middleware.js gzip | 27.9 kB | 27.9 kB | ✓ |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 28.8 kB | 28.8 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 98.4 kB | 98.4 kB | ✓ |
| app-page-tur..prod.js gzip | 99.9 kB | 99.9 kB | ✓ |
| app-page-tur..prod.js gzip | 94.2 kB | 94.2 kB | ✓ |
| app-page.run...dev.js gzip | 157 kB | 157 kB | ✓ |
| app-page.run..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-route-ex...dev.js gzip | 21.4 kB | 21.4 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| server.runti..prod.js gzip | 65.3 kB | 65.3 kB | ✓ |
| Overall change | 975 kB | 975 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | ijjk/next.js add/experimental-traces | Change |
| - | - | - | - |
| 0.pack gzip | 1.61 MB | 1.61 MB | N/A |
| index.pack gzip | 113 kB | 112 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: ea31cfdfc97e99e1c88bb47216c5be8d25200a3a

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts [hunk 1]
```diff
@@ -260,22 +289,60 @@ export class TraceEntryPointsPlugin implements webpack.WebpackPluginInstance {
...
-              nodePath.relative(traceOutputPath, file).replace(/\\/g, '/')
-            )
-          }
+        const readFile = (path: string): Promise<Buffer | string> => {
+          return new Promise((resolve, reject) => {
+            compilation.inputFileSystem.readFile(path, (err, result) => {
+              if (err) {
+                return reject(err)
+              }
+              resolve(result || '')
```
Comment (ztanner) [inline packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:298]: should we treat a missing result as an error too or is it ok for this to potentially return an empty string hash?

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts [hunk 2]
```diff
@@ -260,22 +289,60 @@ export class TraceEntryPointsPlugin implements webpack.WebpackPluginInstance {
...
+            })
+          })
         }
+        const fileHashes: Record<string, string> = {}
+
+        await Promise.all(
```
Comment (ztanner) [inline packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:304]: should we skip this work if `flyingShuttle` is disabled or is there still a use for `fileHashes` outside of it?

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts [hunk 3]
```diff
@@ -294,6 +361,20 @@ export class TraceEntryPointsPlugin implements webpack.WebpackPluginInstance {
...
+    async function getOriginalHash(path: string): Promise<string> {
+      return new Promise((resolve, reject) => {
+        if (path.includes('?')) {
+          path = path.substring(0, path.indexOf('?'))
+        }
+        compilation.inputFileSystem.readFile(path, (err, result) => {
+          if (err) {
+            return reject(err)
+          }
+          resolve(getHash(result || ''))
```
Comment (ztanner) [inline packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:373]: same question about empty string on this one

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts [hunk 4]
```diff
@@ -399,29 +495,35 @@ export class TraceEntryPointsPlugin implements webpack.WebpackPluginInstance {
...
               if (!mod || !mod.dependencies) return
 
               for (const dep of mod.dependencies) {
                 const depMod = getModuleFromDependency(compilation, dep)
 
                 if (depMod?.resource && !depModMap.get(depMod.resource)) {
+                  if (this.flyingShuttle) {
+                    this.traceHashes.set(
+                      depMod.resource,
+                      await getOriginalHash(depMod.resource)
```
Comment (ztanner) [inline packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:508]: should we `Promise.all` these dependency hashes?

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts (hunk 1 above)
Comment (ijjk) [reply packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:298]: I took this as an empty file which is technically valid so don't think we should error

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts (hunk 3 above)
Comment (ijjk) [reply packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:373]: should be the same as https://github.com/vercel/next.js/pull/65071#discussion_r1581301524

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts (hunk 2 above)
Comment (ijjk) [reply packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:304]: We do skip the hashing part if `flyingShuttle` is disabled we still do the `finalFiles` collecting though

File: packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts (hunk 4 above)
Comment (ijjk) [reply packages/next/src/build/webpack/plugins/next-trace-entrypoints-plugin.ts:508]: I didn't for this one as it's recursive and don't want to overload the stack

**Review by ztanner — APPROVED** [review]

**Pull request merged**